services/httpd/handler.go

```diff
@@ -760,6 +777,277 @@ func convertToEpoch(r *query.Result, epoch string) {
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	reqBuf, err := snappy.Decode(nil, compressed)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	var req remote.ReadRequest
+	if err := proto.Unmarshal(reqBuf, &req); err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	if len(req.Queries) == 0 {
+		http.Error(w, "no prometheus queries passed", http.StatusBadRequest)
+		return
+	} else if len(req.Queries) > 1 {
+		http.Error(w, "only one prometheus remote query is supported per request", http.StatusBadRequest)
+		return
+	}
+
+	// Query the DB and create a ReadResponse for Prometheus
+	db := r.FormValue("db")
+	q, err := prometheus.ReadRequestToInfluxQLQuery(&req, db, r.FormValue("rp"))
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	// Check authorization.
+	if h.Config.AuthEnabled {
+		if err := h.QueryAuthorizer.AuthorizeQuery(user, q, db); err != nil {
+			if err, ok := err.(meta.ErrAuthorize); ok {
+				h.Logger.Info(fmt.Sprintf("Unauthorized request | user: %q | query: %q | database %q", err.User, err.Query.String(), err.Database))
+			}
+			h.httpError(w, "error authorizing query: "+err.Error(), http.StatusForbidden)
+			return
+		}
+	}
+
+	opts := query.ExecutionOptions{
+		Database:  db,
+		ChunkSize: DefaultChunkSize,
+		ReadOnly:  true,
+	}
+
+	if h.Config.AuthEnabled {
+		// The current user determines the authorized actions.
+		opts.Authorizer = user
+	} else {
+		// Auth is disabled, so allow everything.
+		opts.Authorizer = query.OpenAuthorizer{}
+	}
+
+	// Make sure if the client disconnects we signal the query to abort
+	var closing chan struct{}
+	closing = make(chan struct{})
+	if notifier, ok := w.(http.CloseNotifier); ok {
+		// CloseNotify() is not guaranteed to send a notification when the query
+		// is closed. Use this channel to signal that the query is finished to
+		// prevent lingering goroutines that may be stuck.
+		done := make(chan struct{})
+		defer close(done)
+
+		notify := notifier.CloseNotify()
+		go func() {
+			// Wait for either the request to finish
+			// or for the client to disconnect
+			select {
+			case <-done:
+			case <-notify:
+				close(closing)
+			}
+		}()
+		opts.AbortCh = done
+	} else {
+		defer close(closing)
+	}
+
+	// Execute query.
+	results := h.QueryExecutor.ExecuteQuery(q, opts, closing)
+
+	resp := &remote.ReadResponse{
+		Results: []*remote.QueryResult{{}},
+	}
+
+	// pull all results from the channel
+	for r := range results {
+		// Ignore nil results.
+		if r == nil {
+			continue
+		}
+
+		// read the series data and convert into Prometheus samples
+		for _, s := range r.Series {
+			ts := &remote.TimeSeries{
+				Labels: prometheus.TagsToLabelPairs(s.Tags),
+			}
+
+			for _, v := range s.Values {
+				timestamp := v[0].(time.Time).UnixNano() / int64(time.Millisecond) / int64(time.Nanosecond)
```

Comment:
`s.Values` is a type from the prometheus package right?  If so, I would be inclined to avoid the unsafe type assertion on elements in `Values`. Only in case a third party bug causes `v[0]` to not be a `time.Time` (which would cause a panic). 

I'd use `ts, ok := v[0].(time.Time)` I think and check `ok`.

One other thing—`time.Nanosecond == 1`. You don't need to divide by it.

Reply:
Good call on the check. On the divide by `time.Nanosecond`, I did that even though it's a 1 because apparently those underlying things aren't guaranteed. Based on the pedantic conversation in the answers of this post: https://stackoverflow.com/questions/24122821/go-golang-time-now-unixnano-convert-to-milliseconds

Reply:
Addressed in f0ffe201615dc9d75c9a6fd064faa076eb99eb16